### PR TITLE
Refactor client game state

### PR DESCRIPTION
**Pull request by nathanleiby**
WIP, but things are mostly behaving well. UI updates feel instantaneous when running multiple clients locally. TBD with internet latency.

Here's the general approach I've taken:

- clients send a variety of typed events to the server. These correspond to user actions
- the server handles the event, updates game state, and runs `syncClientGameState`. This iterates over all of the clients, sending them each a `E.SyncGameState` event that contains their own unique view of the game state.
- in client scenes, `create()` should be safe to do even if `ClientGameState` is empty, just in case the network behavior of syncing the game state comes late. `ClientGameState` should get sync'd on first load, and on subsequent updates
- in client scenes, UI updates no longer occur in the `update()` method, as we only need them to occur on initial load and when `E.SyncGameState` occurs
- (exploratory) instead of passing around the gameState and handling socket updates, use phaser's `this.registry` (a global data store shared across scenes) to persist gameState updates, and subscribe to events from that datastore. 

  
---

One issue I'm trying to resolve is that I keep hitting `Uncaught TypeError: Cannot read property 'cut' of null`:

image

I believe the underlying issue is that that Phaser Scene is getting loaded more than once (e.g. if a new client joins and calls `ServerReady`, it shouldn't cause other clients to re-handle the `ServerReady` behavior in the Preload scene). I've been trying to add additional handling to ensure this cannot occur.

EDIT: It seems like things are *mostly* avoiding this error, but I still hit it sometimes. Needs more debugging. In general, if the UI ever behaves badly, you can safely refresh the page and pull in the latest state.

### CODE REVIEW

**Comment by nathanleiby** [comment]
It's got some rough edges, but things are generally working well and I'd like to merge and iterate from here! 🚢 to heroku and we'll see how it goes there!

LMK what you think or if you have questions as you're working with it. If the UI ever freaks out, just do a page reload and things should happily restore the latest state.

File: src/server/lib/core.ts [hunk 1]
```diff
@@ -1,42 +1,57 @@
...
-    io.to(gameState.room).emit(E.PlayerJoined, <EType[E.PlayerJoined]>{
-      playerID: playerID(socket),
-      playerName: gameState.players.get(playerID(socket)).Name,
-    });
+    gameState.players[id] = {
+      Name: DefaultPlayerName,
+    };
   }
+
+  io.to(socket.id).emit(E.ChangeScene, <EType[E.ChangeScene]>{
```
Comment (peternga) [inline src/server/lib/core.ts:44]: Is this the same as a `socket.emit`?

Comment (nathanleiby) [reply src/server/lib/core.ts:44]: I think so. The goal here is to only call `ChangeScene`  for the newly connected socket.

Above, the goal is to send somthing different to each socket
https://github.com/CleverGameDev/LetterJam/pull/48/files/52b35dbc4307b84d6bc7266184e9d18b41875d22#diff-dcec67ec300a1f50563c2616efebf0ceR22

Comment (nathanleiby) [reply src/server/lib/core.ts:44]: Happy to cleanup things so we're more consistent in how we invoke socket logic.

Comment (nathanleiby) [reply src/server/lib/core.ts:44]: updated to `socket.emit`:
 https://github.com/CleverGameDev/LetterJam/pull/53/commits/6e1d0af97fcacc7d2550e0df2434e18b7fe97a1b

**Pull request merged**